src/lib/RepositoryForms/Form/Processor/Content/UrlRedirectProcessor.php

```diff
@@ -0,0 +1,141 @@
+        array $siteaccessGroups
+    ) {
+        $this->router = $router;
+        $this->requestStack = $requestStack;
+        $this->URLAliasService = $URLAliasService;
+        $this->siteaccessGroups = $siteaccessGroups;
+    }
+
+    /**
+     * @return array
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            RepositoryFormEvents::CONTENT_PUBLISH => ['processRedirectAfterPublish', 0],
+            RepositoryFormEvents::CONTENT_CANCEL => ['processRedirectAfterCancel', 0],
+        ];
+    }
+
+    /**
+     * @param \EzSystems\RepositoryForms\Event\FormActionEvent $event
+     *
+     * @throws \EzSystems\EzPlatformAdminUi\Exception\InvalidArgumentException
+     */
+    public function processRedirectAfterPublish(FormActionEvent $event): void
+    {
+        if (!$this->isAdminSiteAccess()) {
+            return;
+        }
+
+        if ($event->getForm()['redirectUrlAfterPublish']->getData()) {
+            return;
+        }
+
+        /** @var \Symfony\Component\HttpFoundation\RedirectResponse $response */
+        $response = $event->getResponse();
+
+        if (!$response instanceof RedirectResponse) {
+            return;
+        }
+
+        $targetUrlAlias = $this->URLAliasService->lookup(
+            $response->getTargetUrl()
+        );
+
+        $response = new RedirectResponse($this->router->generate(
+            '_ezpublishLocation',
+            ['locationId' => $targetUrlAlias->destination],
+            UrlGeneratorInterface::ABSOLUTE_URL
+        ));
+        $event->setResponse($response);
+    }
+
+    /**
+     * @param \EzSystems\RepositoryForms\Event\FormActionEvent $event
+     *
+     * @throws \EzSystems\EzPlatformAdminUi\Exception\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     */
+    public function processRedirectAfterCancel(FormActionEvent $event): void
+    {
+        if (!$this->isAdminSiteAccess()) {
+            return;
+        }
+
+        /** @var \Symfony\Component\HttpFoundation\RedirectResponse $response */
+        $response = $event->getResponse();
+
+        if (!$response instanceof RedirectResponse) {
+            return;
+        }
+
+        $targetUrlAlias = $this->URLAliasService->lookup(
+            $response->getTargetUrl()
+        );
+
+        $response = new RedirectResponse($this->router->generate(
+            '_ezpublishLocation',
+            ['locationId' => $targetUrlAlias->destination],
+            UrlGeneratorInterface::ABSOLUTE_URL
+        ));
+        $event->setResponse($response);
+    }
+
+    /**
+     * @return bool
+     *
+     * @throws \EzSystems\EzPlatformAdminUi\Exception\InvalidArgumentException
+     */
+    protected function isAdminSiteAccess(): bool
+    {
+        $siteaccess = $this->requestStack->getCurrentRequest()->attributes->get('siteaccess');
```

Comment:
Get current siteaccess from the service `ezpublish.siteaccess`, works with autowiring as well: `\eZ\Publish\Core\MVC\Symfony\SiteAccess`.